app/users/services.py

```
@@ -1,91 +1,101 @@
+        return self.session.query(models.User).filter(models.User.email == email).first()
 
-    new_user = models.User(
-        email=request.email,
-        name=request.name,
-        hashed_password=Hash.bcrypt(request.hashed_password),
-        is_admin=request.is_admin,
-    )
+    def get_users(self):
+        users = self.session.query(models.User).order_by(models.User.id).all()
+        return users
 
-    db.add(new_user)
-    db.commit()
-    db.refresh(new_user)
+    def update_user(self, request: schemas.UserCreate, id: int):
+        user = self.session.query(models.User).filter(models.User.id == id)
 
-    return new_user
+        if not user.first():
+            raise UserNotFoundError
 
+        user.update(request)
+        self.session.commit()
 
-def update(id: int, request: schemas.User, db: Session):
-    user = db.query(models.User).filter(models.User.id == id)
-    if not user.first():
-        raise HTTPException(
-            status_code=status.HTTP_404_NOT_FOUND,
-            detail=f"User with id {id} is not found",
-        )
+        return "Successful Updated"
 
-    user.update(request)
-    db.commit()
+    def get_user(self, id: int):
+        user = self._get_user_by_id(id)
 
-    return "Successful Updated"
+        if not user:
+            raise UserNotFoundError
+        return user
 
+    def delete_user(self, id: int):
+        user = self._get_user_by_id(id)
 
-def get(id: int, db: Session):
-    user = db.query(models.User).get(id)
+        if not user:
+            raise UserNotFoundError
+
+        user.delete(synchronize_session=False)
+        self.session.commit()
+        return "Successful Deleted"
+
+    def authentication(self, request):
+        user = self._get_user_by_email(request.username)
+        if not user:
+            raise InvalidCredentialError
+        if not Hash.verify(
+            request.password,
+            user.hashed_password,
+        ):
+            raise IncorrectPasswordError
```

Hash verify should raise this exception